node/src/types/block.rs

```diff
@@ -891,12 +891,40 @@ impl BlockBody {
         &self.transfer_hashes
     }
 
-    /// Computes the body hash
+    /// Computes the body hash by hashing a Merkle tree of the form:
+    ///                   body_hash
+    ///                   /      \__________________
+    /// hash(deploy_hashes)      /                  \_____________
+    ///                   hash(transfer_hashes)     /             \
+    ///                                         hash(proposer)   hash("")
     pub(crate) fn hash(&self) -> Digest {
-        let serialized_body = self
+        let serialized_deploys = self
+            .deploy_hashes
             .to_bytes()
-            .unwrap_or_else(|error| panic!("should serialize block body: {}", error));
-        hash::hash(&serialized_body)
+            .unwrap_or_else(|error| panic!("should serialize deploy hashes: {}", error));
+        let serialized_transfers = self
+            .transfer_hashes
+            .to_bytes()
+            .unwrap_or_else(|error| panic!("should serialize transfer hashes: {}", error));
+        let serialized_proposer = self
+            .proposer
+            .to_bytes()
+            .unwrap_or_else(|error| panic!("should serialize proposer: {}", error));
+
+        let hashed_deploys = hash::hash(&serialized_deploys).to_array();
+        let hashed_transfers = hash::hash(&serialized_transfers).to_array();
+        let hashed_proposer = hash::hash(&serialized_proposer).to_array();
+        let hashed_empty = hash::hash(&[]).to_array();
+
+        [hashed_deploys, hashed_transfers, hashed_proposer][..]
+            .iter()
+            .rfold(hashed_empty, |prev, next| {
+                let mut to_hash = [0; Digest::LENGTH * 2];
+                to_hash[..Digest::LENGTH].copy_from_slice(&next[..]);
+                to_hash[Digest::LENGTH..].copy_from_slice(&prev[..]);
```

Comment:
Maybe it's a weird question but why are you filling the array "from the end"?

Reply:
In order to preserve the structure of the tree. `rfold` takes elements one by one from the right, and the result of the previous call is the _first_ argument, not the second, so the order appears to be reversed.

For example, the first call will get `(hashed_empty, hashed_proposer)`, and I'm supposed to hash `hashed_proposer + hashed_empty`.

Reply:
I mean, if those two lines were simply reordered? Without changing what gets written where.

Reply:
Ah, but if I understand you correctly, the first line actually fills the first part of the array, and the second line fills the second part. So it's not actually filling "from the end", and that's why it's in this order.

Or do you just prefer the line using `prev` to be the first one?

Reply:
:see_no_evil: :bat: My bad. Sorry. It's OK, just like I like it :)